numpy/core/src/multiarray/ctors.c

```diff
@@ -906,7 +912,7 @@ discover_dimensions(PyObject *obj, int *maxndim, npy_intp *d, int check_it,
         if (PyErr_ExceptionMatches(PyExc_KeyError)) {
```

Comment:
I'm a little confused - does a dictionary actually take this path?

Reply:
changed to `DISCOVERED_OBJECT`

Reply:
Ok, I now understand this now. For dictionary-like objects that implement `__getitem__` but not `__iter__`, this path is taken, as the default `iter(thing_with_getitem)` will call `__getitem__(0)`.

I'd argue that user code that hits this path is itself broken, but that's out of scope for this patch.